### ISSUE TITLE
java.lang.NullPointerException: undefined when using CROSS JOIN UNNEST function

### ISSUE THREAD

**Issue by Chaho12** [prestosql/presto#6235]
Hi. Am am using `CROSS JOIN UNNEST` function as the [document](https://prestosql.io/docs/current/migration/from-hive.html) saids to use instead of `LATER VIEW EXPLODE`.

I am not sure if my UDF is wrong, but it returns `array(varchar)` (in the form of block) successfully (same result as hive)
this is shown when i do show functions
`lago_filter                     | array(varchar)               | array(varchar), varchar                                                        | scalar        | true          |`

What udf_filter do is simply filters using regex expression. 
``` JAVA
   @SqlType("array(varchar)")
    public static Block filterString(@SqlType("array(varchar)") Block areas, @SqlType(StandardTypes.VARCHAR) Slice regex) throws IllegalArgumentException {
        if (areas == null) {
            return null;
        }
        List<String> listAreas = LagoUDFUtils.getListStringFromBlock(areas);
        Pattern pattern = Pattern.compile((regex == null) ? DEFAULT_REGEX : regex.toStringUtf8());
        List<String> filteredArray =  listAreas.stream().filter(collection -> pattern.matcher(collection).matches()).collect(Collectors.toList());

        returnBlock = VARCHAR.createBlockBuilder(null, filteredArray.size());
        filteredArray.forEach(str -> VARCHAR.writeString(returnBlock, str));

        return returnBlock;
    }
}
```
```  SQL
SELECT udf_filter(collection) FROM table1
WHERE log_date = date('2019-01-11')
LIMIT 10;

    _col0
-------------
 []
 [ncs_sdf]
 []
 []
 []
 []
 []
 []
 []
 []
(10 rows)
```

So, the NPE error occurs when I uses this query.
``` SQL
SELECT area, refine_query, count(*) FROM table1
CROSS JOIN UNNEST(udf_filter(collection)) as atf (area)
WHERE log_date = date('2019-01-11')
GROUP BY area, refine_query
LIMIT 10
```

Here is full stack trace shown at web UI.
```
java.lang.NullPointerException: undefined
	at io.prestosql.spi.block.ArrayBlockBuilder.appendStructure(ArrayBlockBuilder.java:150)
	at io.prestosql.spi.type.ArrayType.writeObject(ArrayType.java:189)
	at io.prestosql.$gen.PageProjectionWork_20201207_091554_4.evaluate(Unknown Source)
	at io.prestosql.$gen.PageProjectionWork_20201207_091554_4.process(Unknown Source)
	at io.prestosql.operator.project.DictionaryAwarePageProjection$DictionaryAwarePageProjectionWork.processInternal(DictionaryAwarePageProjection.java:185)
	at io.prestosql.operator.project.DictionaryAwarePageProjection$DictionaryAwarePageProjectionWork.process(DictionaryAwarePageProjection.java:121)
	at io.prestosql.operator.project.PageProcessor$ProjectSelectedPositions.processBatch(PageProcessor.java:333)
	at io.prestosql.operator.project.PageProcessor$ProjectSelectedPositions.process(PageProcessor.java:205)
	at io.prestosql.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:372)
	at io.prestosql.operator.WorkProcessorUtils.lambda$flatten$6(WorkProcessorUtils.java:277)
	at io.prestosql.operator.WorkProcessorUtils$3.process(WorkProcessorUtils.java:319)
	at io.prestosql.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:372)
	at io.prestosql.operator.WorkProcessorUtils$3.process(WorkProcessorUtils.java:306)
	at io.prestosql.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:372)
	at io.prestosql.operator.WorkProcessorUtils.getNextState(WorkProcessorUtils.java:221)
	at io.prestosql.operator.WorkProcessorUtils.lambda$processStateMonitor$2(WorkProcessorUtils.java:200)
	at io.prestosql.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:372)
	at io.prestosql.operator.WorkProcessorUtils.lambda$flatten$6(WorkProcessorUtils.java:277)
	at io.prestosql.operator.WorkProcessorUtils$3.process(WorkProcessorUtils.java:319)
	at io.prestosql.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:372)
	at io.prestosql.operator.WorkProcessorUtils$3.process(WorkProcessorUtils.java:306)
	at io.prestosql.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:372)
	at io.prestosql.operator.WorkProcessorUtils.getNextState(WorkProcessorUtils.java:221)
	at io.prestosql.operator.WorkProcessorUtils.lambda$processStateMonitor$2(WorkProcessorUtils.java:200)
	at io.prestosql.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:372)
	at io.prestosql.operator.WorkProcessorUtils.getNextState(WorkProcessorUtils.java:221)
	at io.prestosql.operator.WorkProcessorUtils.lambda$finishWhen$3(WorkProcessorUtils.java:215)
	at io.prestosql.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:372)
	at io.prestosql.operator.WorkProcessorSourceOperatorAdapter.getOutput(WorkProcessorSourceOperatorAdapter.java:149)
	at io.prestosql.operator.Driver.processInternal(Driver.java:379)
	at io.prestosql.operator.Driver.lambda$processFor$8(Driver.java:283)
	at io.prestosql.operator.Driver.tryWithLock(Driver.java:675)
	at io.prestosql.operator.Driver.processFor(Driver.java:276)
	at io.prestosql.execution.SqlTaskExecution$DriverSplitRunner.processFor(SqlTaskExecution.java:1075)
	at io.prestosql.execution.executor.PrioritizedSplitRunner.process(PrioritizedSplitRunner.java:163)
	at io.prestosql.execution.executor.TaskExecutor$TaskRunner.run(TaskExecutor.java:484)
	at io.prestosql.$gen.Presto_339____20201207_091502_2.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

**Comment by Chaho12** [comment]
Btw. the hive sql below works 
```
SELECT area, refine_query, count(*) FROM table1 
LATERAL VIEW explode(udf_filter(collection)) atf as area 
WHERE log_date = date("2019-01-11")
GROUP BY area, refine_query
LIMIT 10;
```

**Comment by Chaho12** [comment]
Is it alright to not check block size before doing `block.getPositionCount()`?  [line 150](https://github.com/prestosql/presto/blob/master/presto-spi/src/main/java/io/prestosql/spi/block/ArrayBlockBuilder.java#L150) 
Because i check if it is null before doing `getPositionCount()` 
```
if (block != null){
            int blockPC = block.getPositionCount();
```

**Comment by Chaho12** [comment]
In my udf, i have `return null` if input is null for the UDF. I am not too familiar with the concept of `Block` in presto, but should it return `null` or `empty block` when there is `null input`?

``` JAVA
    if (areas == null) {
            return null;
        }
```

**Comment by findepi** [comment]
Why your function has no declaration of `returnBlock` variable type? Is it a local variable of a field in the class?

**Comment by Chaho12** [comment]
@findepi 
ah yes. `private static BlockBuilder returnBlock;`

**Comment by findepi** [comment]
The function can be invoked by multiple times and in multiple threads, so the block builder should be declared locally, inside the funciton.

**Comment by Chaho12** [comment]
Thx. I realized when I was writing test codes and fixed that.

Btw, if I want to return null, do I have to return a block that has null value only?

I know that NPE error happened because I returned `null` for block type.
While I was working a way-around for this error, I realized that if I return `empty block with null value in it` via 
``` JAVA
VARCHAR.createBlockBuilder(null, 0).appendNull().build();
```
 it fixes the NPE error and CROSS JOIN UNNEST works.
